HPCC-19787 Send timestamped signature for Dali file requests

RussWhitehead opened the pull request:
When engines send getPermissions to Dali, send a digital signature of the "username;timestamp"
string and the CDateTime containing that same timestamp. Dali will use that
CDateTime and user to rebuild the string, and call digisign verify() to ensure
the digital signature sent is valid. Additionally, ensure the request did not
hijack a signature by ensuring the timestamp in the signature was within a
configurable threshold. If deemed valid, LDAP authentication is unnecessary.

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

RussWhitehead committed:
HPCC-19787 Send timestamped signature for Dali file requests

When engines send getPermissions() to Dali, send a digital signature of the "username;timestamp"
string and the CDateTime containing that same timestamp. Dali will use that
CDateTime and user to rebuild the string, and call digisign verify() to ensure
the digital signature sent is valid. Additionally, ensure the request did not
hijack a signature by ensuring the timestamp in the signature was within a
configurable threshold. If deemed valid, LDAP authentication is unnecessary.

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexis.com>

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 063fe1c587d11927e77740e9986ea5effefc0230
Number of Error(s): 6
/mnt/disk1/home/vamosax/smoketest/PR-11266/HPCC-Platform/dali/base/dasess.cpp:2125:16: error: cannot allocate an object of abstract type \xe2\x80\x98CUserDescriptor\xe2\x80\x99

[Build error:](https://gist.github.com/415756ec8457ed9d65469c10fc17814a)
Build log [Build log](https://gist.github.com/HPCCSmoketest/415756ec8457ed9d65469c10fc17814a#file-result-18-05-31-18-50-02-log)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 089ea23ad0ff8dd5876935b507fc4f9bc6b09dda
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |786 |786 |0 |
|test (thor) |717 |717 |0 |
|test (roxie) |830 |830 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   15 sec (00:00:15) |  172 sec (00:02:52) |   60 sec (00:01:00) |    7 sec (00:00:07) |   28 sec (00:00:28) | 1200 sec (00:20:00) |   19 sec (00:00:19) | 1501 sec (00:25:01) |

RussWhitehead commented:
@afishbeck Please confirm this design solves the problems we identified
@jakesmith Please review

jakesmith commented on dali/server/daldap.cpp:
```diff
@@ -144,24 +146,50 @@ class CDaliLdapConnection: implements IDaliLdapConnection, public CInterface
+
+        //Check that the digital signature provided by the caller (signature of
+        //caller's "username;timeStamp") matches what we expect it to be
+        if (!isEmptyString(udesc->queryUserTimeStampSignature()))
```
for an 'old' 3.14 this will be just 'signature' , not the expected 'userTimeStampSignature'

RussWhitehead replied:
Same as above, 3.14 is new to Master

jakesmith commented on dali/server/daldap.cpp:
```diff
@@ -144,24 +146,50 @@ class CDaliLdapConnection: implements IDaliLdapConnection, public CInterface
-                StringBuffer b64Signature(udesc->querySignature());
-                if (!pDSM->digiVerify(username, b64Signature))//digital signature valid?
+                StringBuffer requestTimestamp;
+                udesc->queryUTCTimeStamp().getString(requestTimestamp, false);//extract timestamp string from Dali request
```
for 'old' 3.14 clients, no 'utcTimeStamp' will have been deserialized, so requestTimestamp is always going to be 00:00:00...

jakesmith replied:
I see now the this isn't relevant, as won't get to this block if there is no user timestamp signature.

jakesmith commented on dali/server/daldap.cpp:
```diff
@@ -144,24 +146,50 @@ class CDaliLdapConnection: implements IDaliLdapConnection, public CInterface
+
+                CDateTime now;
+                now.setNow();
+                if (now.compare(udesc->queryUTCTimeStamp()) < 0)//timestamp from the future?
```
will be past if 'old' 3.14 afaics, so will deny all old <= 3.14 clients?

RussWhitehead replied:
Older clients will still be sending the legacy style username and password, and this new code will allow it. In this new world, a digisign aware Dali (public/private key files specified in environment) will first verify a signature if it was provided.  If the vierfication fails, the request fails, and if it passee we bypass LDAP authentication.  If no signature is provided in the request, we attempt a LDAP username/pwd authentication.

jakesmith commented on dali/base/dasess.cpp:
```diff
@@ -2055,36 +2066,56 @@ class CUserDescriptor: implements IUserDescriptor, public CInterface
-                mb.read(signature);
+            {
+                mb.read(userTimeStampSignature);
+                if (mb.remaining() > 0)
```
If the serialization format changes, the server version should normally change..
Otherwise you have varying different '3.14' serialization formats.

jakesmith commented:
@RussWhitehead - some questions, main concern is about backward/forward compatibility.
At the moment, I don't see how this works with an older client, and there would be similar issues if new client with old Dali I think.

@afishbeck -tagging you, so on you're list to look also.

RussWhitehead commented on dali/base/dasess.cpp:
```diff
@@ -2055,36 +2066,56 @@ class CUserDescriptor: implements IUserDescriptor, public CInterface
         {
-            if (signature.isEmpty())
+            if (pDSM == nullptr)
+                pDSM = createDigitalSignatureManagerInstanceFromEnv();
```
Will explore renaming queryDigitalSignatureManagerInstanceFromEnv via separate Jira

RussWhitehead replied:
Opened https://track.hpccsystems.com/browse/HPCC-19864 to rename this singleton factory method

RussWhitehead commented on dali/base/dasess.cpp:
```diff
@@ -2055,36 +2066,56 @@ class CUserDescriptor: implements IUserDescriptor, public CInterface
-    void serializeExtra(MemoryBuffer &mb)
+    void serializeSignature(MemoryBuffer &mb)
     {
         if (queryDaliServerVersion().compare("3.14") >= 0)
```
I bumped up to 3.14 during Master (7.0) development, so this will not be a problem. Also, note that this only would be a problem if someone had installed public/private keys and enabled signatures, which so far has only been me

RussWhitehead replied:
3.15 it shall be

RussWhitehead commented on dali/base/dasess.cpp:
```diff
@@ -2055,36 +2066,56 @@ class CUserDescriptor: implements IUserDescriptor, public CInterface
+
+                //Serialize the timestamped signature, and the timestamp object
+                mb.append(userTimeStampSignature.str());
+                now.serialize(mb);
```
I need the time object on the other end to do the time comparisons, which I could recreate from the string. But, I was thinking that the time object is smaller.  If you feel strongly I will change, but it's probably safer to send binary data across the wire since it would be hard to interpret?

RussWhitehead commented:
@jakesmith, please see my comments regarding the Dali interface versioning.  I did make the change to use a local for IDigitalSignatureManager * instead of a member.
@afishbeck You're it

jakesmith commented on dali/base/dasess.cpp:
```diff
@@ -2055,36 +2066,56 @@ class CUserDescriptor: implements IUserDescriptor, public CInterface
+                VStringBuffer toSign("%s;%s", username.str(), timeStr.str());
+
+                StringBuffer userTimeStampSignature;
+                pDSM->digiSign(toSign, userTimeStampSignature);//Sign "username;timeStamp"
```
queryUserTimeStampSignature() and the 'userTimeStampSignature' member are quite confusing because they depend on the context of the IUserDescriptor.
e.g. if you create a IUserDescriptor locally then call queryUserTimeStampSignature() it is undefined.

These members are just part of the internal serialization protocol that now must vary per request, and don't really belong in IUserDescriptor.

For that reason, I think the date signed signature should serialization should be moved outside of IUserDescriptor and instead be part of the request itself.
That way the IUserDescriptor is constant and clean again and you don't have an undefined member and query method depending on context.

i.e. in the client impl. of getPermissionsLDAP, it can call something like : serializeTimeSignedUserDescriptor(IUserDescriptor *user);

And it can perform the same mangling and digi signing.

RussWhitehead replied:
Agreed, will do

RussWhitehead replied:
@jakesmith Actually now I see that I made the timestamp and user;time signature part of the IUserDescriptor so I would not need to add them as arguments to the getPermissionsLDAPmethod, and the ldapConn->getPermissions() method where the actual authentication takes place.
I will pass them as arguments unless maybe I can add generic key/value pair getter/setter to IUserDescriptor and save them that way???   Tony is wanting me to add the scope as part of the signature (scope;user;timestamp) so this would help
Your opion?

jakesmith replied:
It may be convenient to add them the IUserDescriptor implementation(s), but it isn't information that is wanted by users of IUserDescriptor's except by the deserializer and verifier, and the info needs to be different every time (because of date stamp).

So any transient meta info, like date stamp, should be serialized/deserialized when needed and not part of the user descriptor.
If there really need to be extra arguments to getPermissionsLDAP to check LDAP permissions, like date of request, then it is appropriate to add it there.

RussWhitehead replied:
@jakesmith It seems to me that the whole purpose of the IUserDescriptor is to identify and authenticate users.  I can see not serializing/deserializing from the user IUserDescriptor , but it seems that adding generic getProp/setProp is a practical way of adding additional user authentication relics

jakesmith commented:
@RussWhitehead - a few more comments.
Re-tagging @afishbeck as well.

RussWhitehead commented on dali/base/dasess.hpp:
```diff
@@ -77,11 +78,13 @@ interface IUserDescriptor: extends serializable
     virtual const char *querySignature()=0;//user's digital signature
     virtual unsigned querySessionToken()=0;//ESP session token
+    virtual const CDateTime & queryUTCTimeStamp()=0;//Time stamp when a Dali request was requested
+    virtual const char *queryUserTimeStampSignature()=0;//digital signature of "username;TimeStampString"
```
The signature is the signed username, which I create and store once a user has been authenticated via the username;timestamp or LDAP.   I suspect you are correct that it's not used, so I will try to remove it

afishbeck commented:
@RussWhitehead I think the difference between a "token" and a "signature" is that the signature should be tied to the document or message that is being secured.  In this case, if the token is just tied to username and timestamp, it's not really a signature.
I've been looking around to try to find out what "key" and "obj" is in this context but haven't found it.  I haven't switched to looking in the code editor, but presumably the file scope being authorized is there somewhere?  Adding that to the token would make it a true signature, and much harder to reuse in my opinion.

RussWhitehead commented:
@afishbeck Key is the scope, and I will add it to the string being signed. 
So scope;user;timestamp

RussWhitehead committed:
HPCC-19787 Send timestamped signature for Dali file requests

When engines send getPermissions to Dali, send a digital signature of the "username;timestamp"
string and the CDateTime containing that same timestamp. Dali will use that
CDateTime and user to rebuild the string, and call digisign verify() to ensure
the digital signature sent is valid. Additionally, ensure the request did not
hijack a signature by ensuring the timestamp in the signature was within a
configurable threshold. If deemed valid, LDAP authentication is unnecessary.

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexis.com>

RussWhitehead commented:
@jakesmith I implemented the things you requested. Let me know if you prefer it to be squashed first

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 2abf3f592ddea6d5affede9bbed4659035c07a37
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |786 |786 |0 |
|test (thor) |717 |717 |0 |
|test (roxie) |830 |830 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   28 sec (00:00:28) |  170 sec (00:02:50) |   60 sec (00:01:00) |    9 sec (00:00:09) |   29 sec (00:00:29) | 1348 sec (00:22:28) |   19 sec (00:00:19) | 1663 sec (00:27:43) |

jakesmith commented on dali/base/dasess.cpp:
```diff
@@ -453,6 +453,54 @@ class CEnableScopeScansReq : implements IMessageWrapper
+
+    void deserializeSignature(MemoryBuffer &mb, StringBuffer & signature, CDateTime & utcTimeStamp)
+    {
+        if (queryDaliServerVersion().compare("3.15") >= 0)
```
minor: as above, would be clearer/cleaner if version checked by caller

jakesmith commented on dali/base/dasess.cpp:
```diff
@@ -453,6 +453,54 @@ class CEnableScopeScansReq : implements IMessageWrapper
+            if (mb.remaining() > 0)
+            {
+                mb.read(signature);
+                if (mb.remaining() > 0)
```
don't think this is required. If >= 3.15 - will have neither or both signature + datestamp serialized

jakesmith commented on dali/base/dasess.cpp:
```diff
@@ -1405,7 +1462,15 @@ class CCovenSessionManager: public CSessionManagerBase, implements ISessionManag
+    {
+        UNIMPLEMENTED;//client version of getPermissionsLDAP from ISessionManager
+        return SecAccess_Unavailable;
+    }
```
See note where SMSgetPermissionsLDAP is currently called.
I think it should call common getPermissionsLDAP and that method should default the new params to null.
That way, this doesn't have to be unimplemented, i.e. a server-side could still be called (without signature).

jakesmith replied:
@RussWhitehead - I think this one missed, can be removed, if new params are defaulted to null, thus allowing any server caller to call getPermissionsLDAP without a signature.

jakesmith commented on dali/server/daldap.cpp:
```diff
@@ -144,24 +146,50 @@ class CDaliLdapConnection: implements IDaliLdapConnection, public CInterface
         {
             if (nullptr == pDSM)
                 pDSM = createDigitalSignatureManagerInstanceFromEnv();
             if (pDSM && pDSM->isDigiVerifierConfigured())
```
Is it possible for client to have sent 'reqSignature' but for it to be unconfigured here?
(probably is - just checking)

RussWhitehead replied:
Yes it is possible, and unless they also send an LDAP username/pwd the call will fail. You might recall at the breakout meeting in Buckhead last year we discussed this as well as a signature enabled engine sending a request to a legacy Dali, and Richard seemed to agree that a failure was okay

jakesmith replied:
ok

jakesmith commented:
@RussWhitehead - please see new comments/questions.

Tagging @afishbeck - for further review.

RussWhitehead commented:
@jakesmith Implemented your recommendations, please rereview

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 6d417052565f4bb10cae8924d85182fbe02346cc
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |787 |787 |0 |
|test (thor) |718 |718 |0 |
|test (roxie) |831 |831 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   22 sec (00:00:22) |  176 sec (00:02:56) |   60 sec (00:01:00) |    8 sec (00:00:08) |   28 sec (00:00:28) | 1395 sec (00:23:15) |   19 sec (00:00:19) | 1708 sec (00:28:28) |

RussWhitehead commented:
@jakesmith I removed the unimplemented method
@afishbeck I think this is complete, please confirm it suits our needs

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 9181ae6a0a1c6ef28a11f8cf74c7b074724b2fdd
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |787 |787 |0 |
|test (thor) |718 |718 |0 |
|test (roxie) |831 |831 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   23 sec (00:00:23) |  174 sec (00:02:54) |   60 sec (00:01:00) |    7 sec (00:00:07) |   28 sec (00:00:28) | 1215 sec (00:20:15) |   18 sec (00:00:18) | 1525 sec (00:25:25) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 86b82d0c1df495f8b70ec3684917037f3de51a0c
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |787 |787 |0 |
|test (thor) |718 |718 |0 |
|test (roxie) |831 |831 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   18 sec (00:00:18) |  169 sec (00:02:49) |   61 sec (00:01:01) |    7 sec (00:00:07) |   28 sec (00:00:28) | 1406 sec (00:23:26) |   19 sec (00:00:19) | 1708 sec (00:28:28) |

afishbeck commented on dali/base/dasess.cpp:
```diff
@@ -453,6 +453,47 @@ class CEnableScopeScansReq : implements IMessageWrapper
+class CSignatureSerializationHelper
+{
+public:
+    static void serializeSignature(const char * key, IUserDescriptor *udesc, MemoryBuffer &mb)
```
I wish these names were a little more self documenting.  This class name makes me think I could use this class for signatures throughout the code.

Is this really CDaliMessageSignatureHelper, or CDfsScopeSignatureHelper?

How useful / generic is this class?

afishbeck replied:
Yes a more self documenting name would be great.  Also, for example calling "key", "scope" if indeed it is always a scope.  Then I would immediately know when reading the code what was really happening.

RussWhitehead committed:
Signed-off-by: Russ Whitehead <william.whitehead@lexisnexis.com>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 9d54824a441a637a3ff4a72619e8c6265d9a46da
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |787 |787 |0 |
|test (thor) |718 |718 |0 |
|test (roxie) |831 |831 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   34 sec (00:00:34) |  170 sec (00:02:50) |   60 sec (00:01:00) |    9 sec (00:00:09) |   30 sec (00:00:30) | 1333 sec (00:22:13) |   20 sec (00:00:20) | 1656 sec (00:27:36) |

RussWhitehead commented:
@jakesmith @afishbeck Is this ready for merge?

afishbeck reviewed (approved):
@RussWhitehead changes look good.

RussWhitehead commented:
@richardkchapman This is ready for merge

richardkchapman commented:
@jakesmith Are you happy?
